fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -223,23 +152,73 @@
                   @Nonnull IndexStatePrecondition indexStatePrecondition,
                   boolean useSynchronizedSession,
                   long leaseLengthMillis,
-                  boolean trackProgress) {
+                  boolean trackProgress,
+                  @Nonnull IndexFromIndexPolicy indexFromIndex) {
         this.runner = runner;
-        this.recordStoreBuilder = recordStoreBuilder;
         this.index = index;
-        this.recordTypes = recordTypes;
-        this.configLoader = configLoader;
-        this.config = config;
-        this.limit = config.maxLimit;
-        this.syntheticIndex = syntheticIndex;
-        this.indexStatePrecondition = indexStatePrecondition;
-        this.useSynchronizedSession = useSynchronizedSession;
-        this.leaseLengthMills = leaseLengthMillis;
-        this.trackProgress = trackProgress;
-
-        this.recordsRange = computeRecordsRange();
-        timeOfLastProgressLogMillis = System.currentTimeMillis();
-        totalRecordsScanned = new AtomicLong(0);
+        this.indexFromIndex = indexFromIndex;
+
+        this.common = new OnlineIndexerCommon(runner, recordStoreBuilder,
+                index, recordTypes, configLoader, config,
+                syntheticIndex,
+                indexStatePrecondition.getCommonVal(),
+                trackProgress,
+                useSynchronizedSession,
+                leaseLengthMillis
+            );
+    }
+
+    @Nonnull
+    private OnlineIndexerByIndex getScannerByIndex() {
+        if (! (indexScanner instanceof OnlineIndexerByIndex)) { // this covers null pointer
+            indexScanner = new OnlineIndexerByIndex(common, indexFromIndex);
+        }
+        return (OnlineIndexerByIndex)indexScanner;
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> handleScannerReturnOrFallback(Throwable ex, Supplier<CompletableFuture<Void>> fallback) {
+        if (ex == null) {
+            return AsyncUtil.DONE;
+        }
+        if ( indexFromIndex.isActive() &&
+                 indexFromIndex.isAllowRecordScan() &&
+                 (ex.getCause() instanceof OnlineIndexerException)) {
```

Comment:
Maybe it's fine for a draft, but note that we don't want to "fallback" on every exception. It should only fallback on the ones that can possibly be resolved by falling back (e.g. src index is not readable)

Reply:
It is now limited to validation exception: 
`ex.getCause() instanceof OnlineIndexerByIndex.ValidationException`